File: c/baekjoon_17136.cpp
// https://www.acmicpc.net/problem/17136

#include <iostream>
#include <vector>

using namespace std;

int r = 26;
int rest[5] = {0,0,0,0,0};

const inline int min(int a, int b){
    return a < b ? a : b;
}

void search(int x, int y, int size, int d, vector<vector <int>> m) {

    bool check = false;
    for (int i = 0; i < 10; i++){
        for (int j = 0; j < 10; j++){
            if (m[i][j] == 1){
                check = true;
                break;
            }
        }
        if (check){
            break;
        }
    }
    
    if (!check) {
        r = min(r, d-1);
        return;
    }

    if (y+size > 10 || x+size > 10 || d-1 >= r)
        return;
    else {
        bool flag = false;
        for(int i = y; i < y+size; i++) {
            for (int j = x; j < x+size; j++) {
                if (m[i][j]==0){
                    return;
                }
            }
        }
        for(int i = y; i < y+size; i++) {
            for (int j = x; j < x+size; j++) {
                m[i][j] = 0;
            }
        }
    }

    int x2, y2;
    bool flag=false;
    for(y2 =0; y2<10; y2++){
        for (x2 = 0; x2<10; x2++){
            if (m[y2][x2]==1){
                flag = true;
                break;
            }
        }
        if(flag) {
            break;
        }
    }

    for (int i = 5; i > 0; i--) {
        if (rest[i-1] == 5) continue;
        rest[i-1]++;
        search(x2,y2,i,d+1,m);
        rest[i-1]--;
    }
}

int main() {
    vector<vector <int>> m;
    for(int _=0; _<10; _++){
        vector <int> tmp;
        for (int __=0; __<10; __++) {
            int c;
            cin >> c;
            tmp.push_back(c);
        }
        m.push_back(tmp);
    }

    search(0,0,0,0,m);
    
    if (r == 26) {
        cout << -1 << endl;
    }
    else if (r == -1) {
        cout << 0 << endl;

    }
    else {
        cout << r << endl;
    }
}

File: python/baekjoon_17136.py
# https://www.acmicpc.net/problem/17136

global r, rest

def search(x, y, size, d, m):
    global r, rest
    _m = []
    for l in m:
        _m.append(l[:])
    for l in _m:
        if '1' in l:
            break
    else:
        r = min(r, d-1)
        return

    if y+size > 10 or x+size > 10 or x >= 10 or y >= 10 or d-1 >= r:
        return
    else:
        flag = False
        for l in _m[y:y+size]:
            if '0' in l[x:x+size]:
                flag = True
                break
        if flag:
            return
        else:
            for i in range(y, y+size):
                _m[i][x:x+size] = ['0'] * size

    y2=0
    flag=False
    while y2 < 10:
        x2 = 0
        while x2 < 10:
            if _m[y2][x2] == '1':
                flag=True
                break
            x2 += 1
        if flag:    
            break
        y2 += 1

    for i in range(5,0,-1):
        if rest[i-1] == 5:
            continue
        rest[i-1] += 1
        search(x2, y2, i, d+1, _m)  
        rest[i-1] -= 1

def main():
    global r, rest
    m = []
    rest = [0,0,0,0,0]
    r = 26
    for _ in range(10):
        m.append(input().split(' '))

    search(0,0,0,0,m)
    
    if r == 26:
        print(-1)
    elif r == -1:
        print(0)
    else:
        print(r)
    

if __name__ == '__main__':
    main()